src/shadowbox/server/server_access_key.ts

```diff
@@ -90,73 +123,162 @@ export class ServerAccessKeyRepository implements AccessKeyRepository {
-    try {
-      this.keyConfig.write();
-    } catch (error) {
-      throw new Error(`Failed to save config: ${error}`);
-    }
+    this.accessKeys.push(accessKey);
+    this.saveAccessKeys();
     await this.updateServer();
-    return makeAccessKey(this.proxyHostname, accessKeyJson);
+    return accessKey;
   }
 
   removeAccessKey(id: AccessKeyId): boolean {
-    for (let ai = 0; ai < this.keyConfig.data().accessKeys.length; ai++) {
-      const accessKey = this.keyConfig.data().accessKeys[ai];
+    for (let ai = 0; ai < this.accessKeys.length; ai++) {
+      const accessKey = this.accessKeys[ai];
       if (accessKey.id === id) {
-        this.portProvider.freePort(accessKey.port);
-        this.keyConfig.data().accessKeys.splice(ai, 1);
-        this.keyConfig.write();
+        this.portProvider.freePort(accessKey.proxyParams.portNumber);
+        this.accessKeys.splice(ai, 1);
+        this.saveAccessKeys();
         this.updateServer();
         return true;
       }
     }
     return false;
   }
 
-  listAccessKeys(): IterableIterator<AccessKey> {
-    return this.keyConfig.data().accessKeys.map(
-        accessKeyJson => makeAccessKey(this.proxyHostname, accessKeyJson))[Symbol.iterator]();
+  listAccessKeys(): AccessKey[] {
+    // Deep copy the access key array to prevent returning references to keys or their properties.
+    return this.accessKeys.map(key => (key instanceof Object) ? {...key} : key);
   }
 
   renameAccessKey(id: AccessKeyId, name: string): boolean {
-    const accessKeyJson = this.getAccessKey(id);
-    if (!accessKeyJson) {
+    const accessKey = this.getAccessKey(id);
+    if (!accessKey) {
       return false;
     }
-    accessKeyJson.name = name;
+    accessKey.name = name;
     try {
-      this.keyConfig.write();
+      this.saveAccessKeys();
+    } catch (error) {
+      return false;
+    }
+    return true;
+  }
+
+  async setAccessKeyQuota(id: AccessKeyId, quota: AccessKeyQuota): Promise<boolean> {
+    if (!quota || quota.quotaBytes < 1 || quota.windowHours < 1) {
+      return false;
+    }
+    const accessKey = this.getAccessKey(id);
+    if (!accessKey) {
+      return false;
+    }
+    accessKey.quota = quota;
+    try {
+      this.saveAccessKeys();
+      const quotaStausChanged = await this.updateAccessKeyQuotaStaus(accessKey);
+      if (quotaStausChanged) {
+        // Reflect the access key quota status if it changed with the new quota.
+        await this.updateServer();
+      }
+    } catch (error) {
+      return false;
+    }
+    return true;
+  }
+
+  async removeAccessKeyQuota(id: AccessKeyId): Promise<boolean> {
+    const accessKey = this.getAccessKey(id);
+    if (!accessKey) {
+      return false;
+    }
+    const quotaStausChanged = accessKey.isOverQuota;
+    accessKey.quota = undefined;
+    accessKey.isOverQuota = false;
+    try {
+      this.saveAccessKeys();
+      if (quotaStausChanged) {
+        await this.updateServer();
+      }
     } catch (error) {
```

Comment:
I'd like to know a little more about the possible errors here.  Specifically, is it possible and how can our serialized list of AccessKeys get out of sync with the expectations of the manager?  Is it basically just on a file write error?  I guess also more generally, is there a way we expect callers to handle this error other than retrying?  It seems to me like if we can't write a JSON file then there are som ebigger problems happening somewhere

Reply:
`update` can fail due to a write error or a problem starting the shadowsocks process. I've been thinking about this too. IMO, we need typed errors in the access key repository. This way, the caller can take the appropriate actions (i.e. retry, panic).

I decided not to takle this refactor on this PR, since it is already large.

Reply:
Cool.